text/00110-cli-framework-compatibility-strategy.md

```diff
@@ -0,0 +1,464 @@
+We should be able to leverage `jsii` to accomplish this. For example, this is an excerpt from the `.jsii` spec of the `cx-api` module:
+
+```json
+"name": "ContainerImageAssetMetadataEntry",
+"properties": [
+    {
+        "abstract": true,
+        "docs": {
+            "stability": "experimental",
+            "summary": "Path on disk to the asset."
+        },
+        "immutable": true,
+        "locationInModule": {
+            "filename": "lib/assets.ts",
+            "line": 48
+        },
+        "name": "path",
+        "type": {
+            "primitive": "string"
+        }
+    }
+]
+```
+
+This means that, if we remove the `path` property from `ContainerImageAssetMetadataEntry`, jsii should produce an error, and indeed:
+
+```console
+❯ yarn compat
+PROP @aws-cdk/cx-api.ContainerImageAssetMetadataEntry.path: has been removed [removed:@aws-cdk/cx-api.ContainerImageAssetMetadataEntry.path]
+```
+
+Currently, the `cx-api` module contains a lot of stuff that aren't strictly related to the protocol.
+
+> For example, `AssemblyBuildOptions`, which is used during synthesis by the framework. It is not used by the CLI (nor it should).
+
+This creates clutter, and might also cause issues with applying `jsii` compatibility checks.
+
+In addition, according to our new compatibility model, we need to make sure the user gets the correct error
+when the CLI version is smaller than the `Cloud-Assembly` version.
+
+The plan therefore is as follows:
+
+### Step 1: Create a separate package called `cx-protocol`
+
+This package will provide:
+
+#### (1) Object Model
+
+- CloudAssembly
+- RuntimeInfo
+- AssemblyManifest
+- CloudArtifact
+- ...
+- ...
+
+These are the objects that will be later used by the CLI. Running compatibility checks on them **should** ensure:
+
+1. Rename/Remove properties is not allowed.
+2. Changing the type of a property is not allowed.
+3. Making an optional property required is not allowed.
+4. Adding a required property is not allowed.
+
+> The `jsii` compatibility checker currently doesn't actually provide us with all those requirements. It validates interfaces only with relation to their
+> usage, in either argument or return values of methods. This means that, for example, number 4 is allowed,
+> because the `ContainerImageAssetMetadataEntry` is not used as a function argument in `cx-api`.
+> In order to achieve this, we will either need to enhance the `jsii` checker, or refactor our code.
+In any case, we stipulate the necessary changes will be done.
+
+##### (2) De-serialization
+
+The CLI will need to create this object model in-memory, from the disk representation.
+For this, we will provide a de-serialization method:
+
+```typescript
+const assembly = new CloudAssembly("/path/to/cloud-assembly");
+const artifacts = assembly.artifacts;
+const version = assembly.version;
+const runtime = assembly.runtime;
+```
+
+This method will also be used by the framework, as the return value of the `app.synth()` method.
```

Comment:
`app.synth()` *builds* the CloudAssembly in-memory... It doesn't sound natural to me that it'll write files out to disk & than read the assembly from disk... It should create a `CloudAssembly` in-memory, then dump it to disk...

Reply:
> app.synth() builds the CloudAssembly in-memory.

Thats not accurate, it writes to disk as well.

Are you suggesting we change the way we create the cloud assembly directory? Because currently the `CloudAssembly` object is created after all the file system stuff, and i'm pretty sure it will fail if we attempt to instantiate it without the disk representation. 

I agree it would be better to have `CloudAssembly` be purely in-memory, with `save` and `load` methods, thats just not the case right now.

Reply:
Yes, `app.synth` writes to disk and returns a `CloudAssembly` object that provides a high level API for this directory.

Reply:
I've changed this. The new package will not contain the `CloudAssembly` object since we want to leave the new package purely schema. `CloudAssembly` will currently remain in `cx-api`.